### ISSUE TITLE
Detect whether Thrift enveloping is required automatically

### ISSUE THREAD

**Issue by bombela** [yarpc/yab#139]
Now that yarpc disables the thrift envelope by default we should do the same in yab!

**Comment by prashantv** [comment]
Hmm, I'd prefer `yab` work fine with standard HTTP+Thrift outside of YARPC, so I think the approach where it disables for TChannel by default, but not for HTTP makes more sense.

While I don't like that it's not the same as YARPC, I'd like to keep `yab` useful for everyone, not just our specific use.

**Comment by breerly** [comment]
I think I'd want to push back @prashantv - yab should bend in favor of YARPC users, but also be flexible enough to work with Apache Thrift servers. This is going to confuse the heck out of our customers, if they have to disable the envelope for every HTTP+Thrift request.

**Comment by prashantv** [comment]
@breerly we can have a default yabrc internally for our customers which does disable it, the tool itself doesn't need to change for that.

**Comment by prashantv** [comment]
HTTP + Thrift already has a standard (with envelopes), and due to compatibility reasons with TChannel, we made the right decision to simplify for YARPC and always have it disabled by default.

However, that doesn't mean that HTTP+Thrift should be used without envelopes -- I'd suggest that envelopes be enabled by default for HTTP to remain compatible with the rest of the ecosystem.

**Comment by breerly** [comment]
@prashantv agreed - configuring as the default internally sounds like the right tradeoff.

**Comment by breerly** [comment]
@prashantv we've already decided to disable envelopes by default. It was a hard tradeoff to make, but ultimately cleaned up a lot of contentious APIs. Clients can turn on enveloping if they'd like to speak with an Apache Thrift servers, but YARPC+HTTP/1.1 servers will have enveloping off by default. We can jam in person if you feel strongly about that.

**Comment by blampe** [comment]
Could we do some negotiation to figure out whether the envelope should be there or not?

**Comment by jeffbean** [comment]
@blampe It might be a moot point but for benchmarking this might affect measured performance if the process is handling an error on the negotiation? And if warmup=0 and we dont want to conditionalize the testing connections the implicit behavior between warmup and negotiation might be confusing.

**Comment by blampe** [comment]
We can always make benchmarks smarter :)

**Comment by prashantv** [comment]
Closing for now since we have a solution for our internal use case (defaults file).

**Comment by abhinav** [comment]
This is closed but a couple thoughts:

-   IIRC @kriskowal found that there were some ambiguities between the two
    methods which make negotiation of enveloped vs non-enveloped non-trivial.
-   I agree with Prashant that we should continue to work with standard
    Thrift HTTP servers by default.
    
A global internal yabrc that flips this flag sounds reasonable to me.

**Comment by breerly** [comment]
I'd like to reopen this issue and talk about it in person - just burnt another 30 minutes wondering wth was wrong with my server.

**Comment by breerly** [comment]
Ok - *sounds* like the conclusion is that we can make yab detect whether or not to do enveloping.

**Comment by prashantv** [comment]
@abhinav @breerly 

I took a stab at this, but I don't think auto detection is possible due to the yarpc-go's implementation of bad requests.

WIP branch:
https://github.com/yarpc/yab/compare/bump_deps...detect_thrift?expand=1
```
case YARPC HTTP (enveloped) false 0
Call got response <nil> HTTP call got non-success response code: 400, body: BadRequest: unrecognized procedure "__INVALID_SVC::__INVALID_METHOD" for service "foo"

case YARPC HTTP (non-enveloped) true 0
Call got response <nil> HTTP call got non-success response code: 400, body: BadRequest: unrecognized procedure "__INVALID_SVC::__INVALID_METHOD" for service "foo"
```

I send an enveloped request with a method name that shouldn't be found. An Apache Thrift server responds with an enveloped exception reply. YARPC always responds with a 400 rather than an enveloped reply, even when I explicitly enable envelopes in YARPC.

Couple of other options:
 - Is it possible for HTTP inbound to respond with an enveloped exception on unknown methods (similar to Apache Thrift)? I'm guessing not, since the dispatcher can't tell what encoding the response should be when the method isn't found.
 - Should we be using separate `RPC-Encoding` values for enveloped vs unenveloped requests, and have the Thrift registrations in YARPC support both enveloped and unenveloped?

I'm leaning towards the latter, it really feels like encoding should uniquely describe the protocol used to encode the bytes, and enveloped Thrift is really a different protocol.

**Comment by abhinav** [comment]
I sort of assumed that auto-detection was for old Apache Thrift services only. I like the option of YARPC services registering both, enveloped and unenveloped and dispatching on the encoding (or a different header?).

**Comment by abhinav** [comment]
I sort of assumed that auto-detection was for old Apache Thrift services only. I like the option of YARPC services registering both, enveloped and unenveloped and dispatching on the encoding (or a different header?).

**Comment by breerly** [comment]
@abhinav I'm actually open to that. We can default to non-enveloped, but in the presence of a certain header we can dispatcher on the enveloped encoding.

**Comment by breerly** [comment]
The other options is to treat enveloped as a separate encoding - `RPC-Encoding: thrift-enveloped`.

**Comment by prashantv** [comment]
@abhinav Since YARPC services support enveloping, the auto-detection has to support both Apache Thrift services as well as YARPC services.

I do think dispatching on encoding is the right solution, but I'm not sure if that would be strictly backwards compatible since:
 - A yarpc client/handler with enveloping uses `thrift` as the encoding.
 - If we wanted to dispatch on some new encoding (`thrift-enveloped`), old clients would not be compatible with new services (and vice-versa).

Unenveloped YARPC clients/services (the default) would be completely backwards compatible.

**Comment by Willyham** [comment]
I'd like to reopen this discussion. I just spent the best part of 3 hours debugging what was wrong with my code and it turned out all I needed was the `--disable-thrift-envelope` flag on my request (I'm working outside of the Uber ecosystem here).

At the very least we need to add some solid documentation around this. I followed the yarpc docs to set up a thrift service with an HTTP inbound and then tried to use `yab` to test it. I feel like this is a pretty standard use case that people will encounter straight off the bat.

The error it fails with is extremely difficult to debug:

```
yab -t ./idl/foo.thrift -p "http://localhost:8080/rpc" foo Foo::Test -r '{"request":"test"}'
Failed while making call: HTTP call got non-success response code: 400, body: BadRequest: failed to decode "thrift" request body for procedure "Foo::Test" of service "foo" from caller "yab-willy": unknown ttype Type(128)
```

I had to strip my thrift file down to the absolute minimum before I realised that this wasn't a problem actually decoding one of the types that I'd defined. At that point I got desperate and trawled through these issues before finding the solution.

Can we consider detecting this case and suggesting the `--disable-thrift-envelope` flag in the error message?

**Comment by blampe** [comment]
Proposal in #741.

**Comment by kevinguard** [comment]
@Willyham You saved my day! `--disable-thrift-envelope` did the job!

**Comment by breerly** [comment]
We're looking to release auto-enveloping in YARPC [v1.28.0](https://github.com/yarpc/yarpc-go/blob/dev/CHANGELOG.md#unreleased), this will make requests to YARPC HTTP servers work *regardless* of whether `--disable-thrift-envelope` is passed!

Once that is out, this ticket can be closed.

**Comment by blampe** [comment]
Released in v1.28.0.
